File: frontend/types/basket.ts
import type { DeliveryAddressRead, DeliveryRecipientRead } from "@/services/api/order-drafts.types"

export type BasketItemCreate = {
    variant_id: number
    quantity: number
}

export type BasketItemUpdate = {
    quantity: number
}

export type BasketProductSummaryRead = {
    id: number
    sku: string
    name: string
    in_stock: boolean
    image_url: string
}

export type BasketVariantSummaryRead = {
    id: number
    sku: string | null
    name: string
    stock: number
    price: string
    image_url: string
}

export type BasketItemRead = {
    id: number
    variant_id: number
    quantity: number
    unit_price: string
    line_total: string
    available_quantity: number
    is_available: boolean
    product: BasketProductSummaryRead
    variant: BasketVariantSummaryRead
    created_at: string
    updated_at: string
}

export type BasketRead = {
    id: number
    user_id: number
    items: BasketItemRead[]
    delivery_address_id: number | null
    recipient_id: number | null
    delivery_address: DeliveryAddressRead | null
    recipient: DeliveryRecipientRead | null
    items_count: number
    total_quantity: number
    total_amount: string
    delivery_total: string
    grand_total: string
    currency: string
    delivery_period_min: number | null
    delivery_period_max: number | null
    has_unavailable_items: boolean
    created_at: string
    updated_at: string
}

File: backend/src/database/schemas/basket/basket.py
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field

from src.database.schemas.basket.basket_item import BasketItemRead
from src.database.schemas.delivery.address import DeliveryAddressRead
from src.database.schemas.delivery.recipient import DeliveryRecipientRead


class BasketBase(BaseModel):
    user_id: int = Field(ge=1)


class BasketCreate(BasketBase):
    pass


class BasketUpdate(BaseModel):
    user_id: int | None = Field(default=None, ge=1)
    delivery_address_id: int | None = Field(default=None, ge=1)
    recipient_id: int | None = Field(default=None, ge=1)
    delivery_total: Decimal | None = Field(default=None, ge=0, max_digits=12, decimal_places=2)
    currency: str | None = None
    delivery_period_min: int | None = Field(default=None, ge=0)
    delivery_period_max: int | None = Field(default=None, ge=0)


class BasketRead(BasketBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    items: list[BasketItemRead]
    delivery_address_id: int | None = None
    recipient_id: int | None = None
    delivery_address: DeliveryAddressRead | None = None
    recipient: DeliveryRecipientRead | None = None
    items_count: int = Field(ge=0)
    total_quantity: int = Field(ge=0)
    total_amount: Decimal = Field(ge=0, max_digits=12, decimal_places=2)
    delivery_total: Decimal = Field(ge=0, max_digits=12, decimal_places=2)
    grand_total: Decimal = Field(ge=0, max_digits=12, decimal_places=2)
    currency: str
    delivery_period_min: int | None = None
    delivery_period_max: int | None = None
    has_unavailable_items: bool
    created_at: datetime
    updated_at: datetime
